### ISSUE TITLE
New ✨ Implement Bloom Filter for Bluedis

### ISSUE THREAD

**Issue by IAmRiteshKoushik** [IAmRiteshKoushik/bluedis#12]
## Issue Description

The issue seeks the implementation of the probabilistic data structure `Bloom Filter` which is used to check for the presence of an item. 

Here's a good read - [Bloom Filters Explained](https://systemdesign.one/bloom-filters-explained/)
Here's the redis docs on the same - [Bloom Filters in Redis](https://redis.io/docs/latest/develop/data-types/probabilistic/bloom-filter/)

## High Level Goals
- [ ] Implement Bloom Filter Scheme

> [!NOTE]
> As this is a systems programming project, issue are not assigned on a first-come-first-serve basis but rather based on the best approach shared by the contributors for implementing the mentioned feature. Please share your approach below!

**Comment by yendelevium** [comment]
Hey @IAmRiteshKoushik , I'm interested in taking up this issue. I think it's pretty straight forward after reading the docs, but I'm not sure about how big the size should be. If i use too much, it might be a waste of space, and if I use too little it will increase the false negative rate. Any pointers on this? Apart from that, I'm thinking of using 3-4 hash functions and  a slice/array to keep track of the bits(for constant indexing) Concurrency safety can be managed by RWMutex (As I'm assuming we will be reading a lot more than writing after a point in time)

**Comment by IAmRiteshKoushik** [comment]
> Hey @IAmRiteshKoushik , I'm interested in taking up this issue. I think it's pretty straight forward after reading the docs, but I'm not sure about how big the size should be. If i use too much, it might be a waste of space, and if I use too little it will increase the false negative rate. Any pointers on this? Apart from that, I'm thinking of using 3-4 hash functions and a slice/array to keep track of the bits(for constant indexing) Concurrency safety can be managed by RWMutex (As I'm assuming we will be reading a lot more than writing after a point in time)

Hi @yendelevium . I was thinking of capping the max size at 256MB. Can you list out the hash functions that you are planning to use along with their requirement ? 

If this is your first time coding a Bloom Filter then please check out [LIVE Coding BloomFilter with Arpit Bhayani](https://www.youtube.com/watch?v=UVFnabieyzc&t=696s&pp=ygUhYXJwaXQgYmhheWFuaSBibG9vbSBmaWx0ZXIgc2NoZW1l). Try implementing this before taking up this issue ?

**Comment by yendelevium** [comment]
> > Hey @IAmRiteshKoushik , I'm interested in taking up this issue. I think it's pretty straight forward after reading the docs, but I'm not sure about how big the size should be. If i use too much, it might be a waste of space, and if I use too little it will increase the false negative rate. Any pointers on this? Apart from that, I'm thinking of using 3-4 hash functions and a slice/array to keep track of the bits(for constant indexing) Concurrency safety can be managed by RWMutex (As I'm assuming we will be reading a lot more than writing after a point in time)
> 
> Hi @yendelevium . I was thinking of capping the max size at 256MB. Can you list out the hash functions that you are planning to use along with their requirement ? 
> 
> If this is your first time coding a Bloom Filter then please check out [LIVE Coding BloomFilter with Arpit Bhayani](https://www.youtube.com/watch?v=UVFnabieyzc&t=696s&pp=ygUhYXJwaXQgYmhheWFuaSBibG9vbSBmaWx0ZXIgc2NoZW1l). Try implementing this before taking up this issue ? 

Yes, it is my first time coding a bloom filter. I will implement it and get back to you.m

**Comment by yendelevium** [comment]
Hey @IAmRiteshKoushik , here's how i plan to implement the bloom filter
The hashes i will be using
- FNV hash
- xxHash
- Murumurhash
These are non-cryptographic hashes, so they are very fast. These specific hashes also have a comparatively lesser collision rate than other fast hashes, so this will help ensure that the bloomfilter is fast while reducing the chances of false positives

Next, the size of the filter. Generally, a bool can be used to indicate whether the bucket's bit is set to 1 or 0. We get around 256,000,000 buckets. But this wastes 7 bits. So, instead of using bool, we can use a unsigned integer, and change its respective bits to 1/0.  This lets us have 2,048,000,000 buckets, reducing the collision rate immensely so the no.of false positives decreases as well. But tbh i think this is overkill so I would probably reduce the size to around 1,000,000,000 (125MB). But i can proceed with 256MB as well.

If this is fine, can I take up this issue?

**Comment by IAmRiteshKoushik** [comment]
Sounds good @yendelevium. Assigning this issue to you. Make sure to also checkout how redis-cli interacts with bloom filters and drop in the corresponding supporting commands to parse and perform the actions.

**Comment by yendelevium** [comment]
Sure
